### PR TITLE
[SPARK-33874][K8S] Handle long lived sidecars

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

For liveness check when checkAllContainers is not set, we check the liveness status of the Spark container if we can find it.

### Why are the changes needed?

Some environments may deploy long lived logs collecting side cars which outlive the Spark application. Just because they remain alive does not mean the Spark executor should keep running.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Extended the existing pod status tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Thanks @holdenk for cc'ing me. I'll try to run tests locally and take a look tonight in KST. cc @dongjoon-hyun too FYI.

**Review by HyukjinKwon — APPROVED** [review]
Looks making sense to me.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala:101]: Maybe?
```scala
val sparkContainerName = sc.conf.get(KUBERNETES_EXECUTOR_PODTEMPLATE_CONTAINER_NAME)
  .getOrElse(DEFAULT_EXECUTOR_CONTAINER_NAME)
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala:104]: In this case, one liner will be better.
```scala
ExecutorPodsSnapshot.setSparkContainerName(sparkContainerName)
```

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala:119]: indentation?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala:112]: This looks like a test helper because this is used in the test case only. In that case, shall we move this to `ExecutorPodsSnapshotSuite.scala`?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:98]: When does this happen? This looks like the case, `pod.getStatus.getContainerStatuses.asScala.find(_.getName() == sparkContainerName) == None`. In that case, isn't it a bug?

Comment (HyukjinKwon) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:98]: Yeah, I misread the codes. This looks it should be fixed. I thought `spark.kubernetes.executor.podTemplateContainerName` is optional so it's fine but I just realised that we're setting a default value below.

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:98]: So the idea here is that if for some reason we can't find the Spark container defaulting to the pod status is the most reasonable thing to do. The alternative would be to throw an exception here since it indicates something unexpected has happened.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:90]: We can also have the situation where there is no terminated state, so it should be pod running, but I can collapse it into a single case/switch.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala
Comment (holdenk) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala:112]: I don't have a strong preference, this is the other place where the utils functions for `ExecutorPodsSnapshotSuite` live though so I was going for consistency.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:98]: In that case, `logWarning` will be more helpful for us to detect it and fix it.

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:98]: Sounds good, I'll add a logWarning to this.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala:112]: Got it.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
> Sounds good, I'll add a logWarning to this.

@holdenk . According to the error message, the last commit looks wrong to me because it added `logWarning` at line 94 instead of line 99.

Could you (1) revise the error message of line 94 properly and (2) add a new `logWarning` at line 99 of `ExecutorPodsSnapshot.scala`, please?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:94]: Shall we revise this because we already find a corresponding container in this code path via `Some(sparkContainerStatus)` while we don't know the termination code?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:101]: Indentation?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating, @holdenk .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (except two new minor comments).

Please feel free to merge this after addressing the above two.

**Comment by holdenk** [comment]
Thanks @dongjoon-hyun I'll clean that up today :)

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.1 and branch-3.0.

**Comment by HyukjinKwon** [comment]
@holdenk, should we port this back to branch-3.0 too? It has a conflict. I will leave it to you.

**Comment by holdenk** [comment]
Thanks, I think we can leave it out of 3.0, supporting sidecars is relatively new our the code base.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:85]: Sorry, but this seems to fail like the following. Could you check this once more, @holdenk ?
```
containerStatuses=[
ContainerStatus(
  containerID=docker://...,
  image=...,
  imageID=..., 
  lastState=ContainerState(running=null, terminated=null, waiting=null, additionalProperties={}),
  name=pod,
  ready=true,
  restartCount=0,
  started=true,
  state=ContainerState(running=ContainerStateRunning(startedAt=2021-01-03T10:25:24Z, additionalProperties={}), terminated=null, waiting=null, additionalProperties={}),
  additionalProperties={}
)
]
```

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:85]: That makes sense, we'll sometimes not have any status entries for the Spark container. I'll take out the logWarning since this is a more expected state.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:85]: Thank you, @holdenk .

**Pull request closed without merging**
